napari/layers/utils/layer_utils.py

```diff
@@ -208,10 +212,119 @@ def convert_to_uint8(data: np.ndarray) -> np.ndarray:
             ).astype(out_dtype)
 
 
+def prepare_properties(
+    properties: Optional[Union[Dict[str, Array], DataFrame]],
+    choices: Optional[Dict[str, Array]] = None,
+    num_data: int = 0,
+    save_choices: bool = False,
+) -> Tuple[Dict[str, np.ndarray], Dict[str, np.ndarray]]:
+    """Prepare properties and choices into standard forms.
+
+    Parameters
+    ----------
+    properties : dict[str, Array] or DataFrame
+        The property values.
+    choices: dict[str, Array]
+        The property value choices.
+    num_data: int
+        The length of data that the properties represent (e.g. number of points).
+    save_choices: bool
+        If true, always return all of the properties in choices.
+
+    Returns
+    -------
+    properties: dict[str, np.ndarray]
+        A dictionary where the key is the property name and the value
+        is an ndarray of property values.
+    choices: dict[str, np.ndarray]
+        A dictionary where the key is the property name and the value
+        is an ndarray of unique property value choices.
+    """
+    # If there is no data, non-empty properties represent choices as a deprecated behavior.
+    if num_data == 0 and properties:
+        warnings.warn(
+            trans._(
+                "Property choices should be passed as property_choices, not properties. This warning will become an error in version 0.4.11.",
+            ),
+            DeprecationWarning,
+            stacklevel=2,
+        )
+        choices = properties
+        properties = {}
+
+    properties = validate_properties(properties, expected_len=num_data)
+    choices = _validate_property_choices(choices)
+
+    # Populate the new choices by using the property keys and merging the
+    # corresponding unique property and choices values.
+    new_choices = {
+        k: np.unique(np.concatenate((v, choices.get(k, []))))
+        for k, v in properties.items()
```

Comment:
The use of `np.unique` here and in other places in this file is probably what causes #2756. Is this something we can get around at all?

Reply:
Thanks for the pointer to #2756. That issue is specifically caused by lack of support for `dtype=object` when using `np.unique` with the `axis` argument. I think that's related to how we sometimes store a collection of colors as an Nx4 numpy array, so using `axis=0` in those cases is safe, but using it on more general properties is not. More motivation to unify properties I suppose!

Here, we're not using `axis`, but may run into similar problems when a properties array value contains mixed types or contains None (e.g. to represent a missing property value). Also, I feel like this may not be behaving as intended (before and after this PR) if `properties` passed in does contain color array values in the Nx4 form mentioned above. That's probably why a lot of color properties are treated as a special cases - I think we either need to completely separate those from more general properties or find a way a unify them to avoid problems like this.

Reply:
One possible solution is making a stronger/more explicit of a Property type (i.e. not just a name and an array-like). Also consider that Pandas does handle some of these problems more gracefully/usefully/efficiently than numpy (e.g. better handling of None to represent missing data).